### PR TITLE
add thinning

### PR DESCRIPTION
**Pull request by StevenPuttemans**
New functionality to thin out blobs if desired. Used it to perform skeletization on segmented blobs in fingerprints, to find minutae (intersections). Example of application can be seen below:

### CODE REVIEW

File: modules/ximgproc/src/thinning.cpp [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+namespace cv {
+namespace ximgproc {
+
+// Function prototypes to suppress -Wmissing-declarations warnings
+void thinningIteration(Mat img, int iter);
+void thinning(InputArray _input, OutputArray _output);
+
+// Applies a thinning iteration to a binary image
+void thinningIteration(Mat img, int iter){
```
Comment (alalek) [inline modules/ximgproc/src/thinning.cpp]: Use "static" here.
Remove all these function prototypes.
Add include "opencv2/ximgproc.hpp"

Comment (StevenPuttemans) [inline modules/ximgproc/src/thinning.cpp]: Okay will do so! Could you explain why the Ni Black hasn't got the ximgproc include but doesnt trigger an issue with the -Wmissing declarations?

Comment (alalek) [inline modules/ximgproc/src/thinning.cpp]: Well, additional include is not required. It is done in [precomp.hpp file](https://github.com/opencv/opencv_contrib/blob/master/modules/ximgproc/src/precomp.hpp)

File: modules/ximgproc/include/opencv2/ximgproc.hpp [hunk 2]
```diff
@@ -104,6 +104,16 @@ CV_EXPORTS_W void niBlackThreshold( InputArray _src, OutputArray _dst,
...
                                     int blockSize, double delta );
 
+/** @brief Applies a binary blob thinning operation, to achieve a skeletization of the input image.
+
+The function transforms a binary blob image into a skeletized form using the technique of Zhang-Suen.
+
+@param _src Source 8-bit single-channel image, containing binary blobs, with blobs having 255 pixel values.
+@param _dst Destination image of the same size and the same type as src. The function can work in-place.
+ */
+CV_EXPORTS_W void thinning( InputArray _src, OutputArray _dst);
```
Comment (alalek) [inline modules/ximgproc/include/opencv2/ximgproc.hpp]: Could you please remove underscore "_" from parameters here and from documentation comment.

File: modules/ximgproc/src/thinning.cpp (hunk 1 above)
Comment (StevenPuttemans) [inline modules/ximgproc/src/thinning.cpp]: Well but if i do not add the prototypes then I get this error when building locally. Will see what buildbot says.

File: modules/ximgproc/include/opencv2/ximgproc.hpp (hunk 2 above)
Comment (StevenPuttemans) [inline modules/ximgproc/include/opencv2/ximgproc.hpp]: Will do so!

**Comment by alalek** [comment]
Thank you! :+1:

**Comment by StevenPuttemans** [comment]
Thank you!

**Pull request merged**